### ISSUE TITLE
Trying to understand mapping callbacks

### ISSUE THREAD

**Issue by bandokihiro** [StanfordLegion/legion#1220]
I am attempting to enforce a given mapping order. To do this, I need to hold onto a task in `select_tasks_to_map`. Trying to implement this, I have a few questions regarding the order of mapping callbacks in general. I am starting with a single rank run with 1 sub-region (i.e. no partitioniong). I create one mapper per process (there is only one process running at the moment). I override the default mapper behavior but, for now, none of the default behavior is modified. I am printing stuff in `select_task_options`, `select_tasks_to_map`, `slice_task`, and `map_task` for a particular task. This is a sample output
```
Selecting options for volume_residual_task; unique ID = 376
Selecting options for volume_residual_task; unique ID = 381
Selecting options for volume_residual_task; unique ID = 386
Inserting volume_residual_task; unique ID = 403
Slicing volume_residual_task; unique ID = 403
Inserting volume_residual_task; unique ID = 404
Mapping volume_residual_task; unique ID = 405

Inserting volume_residual_task; unique ID = 418
Slicing volume_residual_task; unique ID = 418
Inserting volume_residual_task; unique ID = 419
Mapping volume_residual_task; unique ID = 420

Inserting volume_residual_task; unique ID = 433
Slicing volume_residual_task; unique ID = 433
Inserting volume_residual_task; unique ID = 434
Mapping volume_residual_task; unique ID = 435
```
The unique ID in a legion prof picture always corresponds to the one I see in map task. My questions are:

1. why is `select_task_options` called 3 times first? Each iteration in my application is made of 3 stages (an actual algorithmic iteration). The traced region is an iteration/set of 3 stages  but this output comes from a run without `-dm:memoize`
2. why is `select_task_options` called twice, once before `slice_task` and once after it? Does is get called the first time after sharding but before slicing?
3. if so, how can I make a difference between the two occurences? What I need to do is put this task into the pipeline for `map_task` after I made sure others are already in there.

Thank you for your help.

**Comment by lightsighter** [comment]
The runtime executes tasks like a pipelined out-of-order processor. Each stage operates independently of the others and tasks don't have to go all the way through the pipeline before the next task can start the pipeline, even in the presence of data dependences. Pipeline stages are unbalanced in the amount of work they each require and by the amount of work required for each individual task or operation within a stage. Furthermore, the stages are serviced by a limited number of utility workers, usually less than the number of stages. The runtime has a number of scheduling heuristics that attempt to balance throughput of tasks moving through the pipeline with the latency that any one task experiences. We also try to keep pipeline stages "full" enough so that bubbles aren't introduced in the pipeline. In general, I wouldn't try to reason about these heuristics and how they work, especially since they are liable to change over time. Pretty much the only ordering assumptions about callbacks you can guarantee are about the ordering of callbacks for the same operation. I believe @manopapad has a diagram that shows the permitted order of callbacks for a given operation traversing the pipeline. There are few different reorderings of mapper calls depending on the output of earlier mapping calls, but in general there aren't too many possibilities. All that said, now we can try to answer your questions.

1. The `select_task_options` call occurs in the "prefix" stage of the pipeline and can be done in any order for any operation after it is launched. It's in expensive and the runtime is prioritizing the execution of earlier stages in the pipeline to try and fill it up because the later stages are always empty at the beginning and it perceives those as "bubbles" so it's trying to eliminate them.
2. Where do you see `select_task_options` being called twice? I see exactly one invocation for each of the three tasks. Perhaps you meant `select_tasks_to_map`. If so, the answer really depends on your output of `slice_task`. What did you set the output of `recurse` to in you `SliceTaskOutput`?
https://gitlab.com/StanfordLegion/legion/-/blob/master/runtime/legion/legion_mapping.h#L446 
3. Probably need to answer my question above before we can tackle this one.

**Comment by bandokihiro** [comment]
> Pretty much the only ordering assumptions about callbacks you can guarantee are about the ordering of callbacks for the same operation

Here, by operation, do you mean an app task for instance?

@manopapad I would be very interested in such diagram if you have one.

> Perhaps you meant `select_tasks_to_map`

Yes, sorry for the confusion. I use the default `slice_task` implementation. Previously, I did override it but I realized that the same behavior I was trying to enforce could simply by obtained with `-dm:same_address_space` so that's what I am doing now.

**Comment by manopapad** [comment]
I never made a real diagram, but here's an overview:

* `select_task_options` (once per task launch)
* `select_tasks_to_map` (until task gets added to list)
* `select_sharding_functor` (once per task launch, if replicated execution)
* `slice_task` (if index launch, can recurse, includes only local points if replicated execution)
* `map_task`
* `select_task_sources` (for each region requirement, to decide where to pull the data from, if more than one option)

I suggest you also try using the `LoggingWrapper` when debugging mappers: https://github.com/StanfordLegion/legion/blob/master/runtime/mappers/logging_wrapper.h

**Comment by bandokihiro** [comment]
do you see a reason why `select_tasks_to_map` would be called twice for a single task launch?

**Comment by manopapad** [comment]
It is possible that the first call is referring to the index launch as a whole, and the subsequent calls are handling each point task individually after slicing (in this case there's only one point task). I am not certain about this part of the workflow, because @lightsighter and @rohany were making changes to it recently.

**Comment by rohany** [comment]
> do you see a reason why select_tasks_to_map would be called twice for a single task launch?

As mike mentioned above, I believe that this only happens if you set `recurse` to true in your slice task.

**Comment by bandokihiro** [comment]
I don't think the default mapper does that?

**Comment by manopapad** [comment]
We can easily test the theory that `select_tasks_to_map` is getting called once for the overall launch and once for each point task. Just do a larger index launch, with 2 point tasks. If the theory is correct, then you should see 3 total instances of `select_tasks_to_map`.

**Comment by bandokihiro** [comment]
It looks insensitive to the size of the launch space.

**Comment by lightsighter** [comment]
The only tasks that are end up in `select_tasks_to_map` are either individual tasks launched by `IndividualTaskLauncher` or slice tasks that came from a previous call to `slice_task` for an index space task launched by `IndexTaskLauncher`. If you're running with control replication, then the behavior you're observing is expected because sharding is an implicit slicing. The process of sharding any index task will always produce a slice task with `recurse=true` so that the mapper always gets a chance to further slice the local set of point tasks of the index task. The runtime calls `select_tasks_to_map` to give the mapper an opportunity to move the slice coming out of the sharding process before it calls `slice_task` on it since it doesn't have that ability during the sharding process. After that you get the call to `slice_task` to perform the slicing of the local point tasks, and then the resulting slice(s) goes/go back into the ready queue for `select_tasks_to_map`.

**Comment by bandokihiro** [comment]
OK, this makes sense. In my case, the sharding distributes N point tasks to N ranks and I guess `slice_task` becomes a no-op, i.e. both calls to `select_tasks_to_map` actually see point tasks or, in other words, `slice_task` doesn't slice, only the sharding does. Is there a way for me to hold onto a point task at the second call to `select_tasks_to_map`, i.e. "right before" it falls into `map_task`?

**Comment by lightsighter** [comment]
> In my case, the sharding distributes N point tasks to N ranks and I guess slice_task becomes a no-op, i.e. both calls to select_tasks_to_map actually see point tasks or, in other words, slice_task doesn't slice, only the sharding does.

If people feel like it would be useful, we could extend the `select_sharding_function` call to say whether the resulting slice should be considered a `recurse` slice or just a slice which does not need further slicing. It wouldn't be a hard thing to do.

> Is there a way for me to hold onto a point task at the second call to select_tasks_to_map, i.e. "right before" it falls into map_task?

This is what @rohany wanted to do as well. Since you control the granularity of slicing, you can continue slicing all the way down to individual points if you want to, and then prevent slices containing just a single point from moving to `map_task` by keeping them in the ready queue for `select_tasks_to_map`.

**Comment by bandokihiro** [comment]
So I think, at least for the N points to N ranks case, that both calls to `select_tasks_to_map` (the one after sharding and the one after slicing) will see single point tasks (the same one beside the fact that the unique ID seems to get incremented between these two stages).

**Comment by lightsighter** [comment]
> So I think, at least for the N points to N ranks case, that both calls to select_tasks_to_map (the one after sharding and the one after slicing) will see single point tasks

We should probably provide a way for mappers to say that the result of sharding does not need to be further sliced.

> (the same one besides the fact that the unique ID seems to get incremented between these two stages).

Right, because conceptually you made a new slice. Each call to `slice_task` will make completely new slices and throw away the old one. We presume that if you call `slice_task`, then you're probably going to slice it in some way.

> And I think I was holding onto both (or something in that flavor) which caused a hang.

Yes, if you thought you still had to do something with the original slice task, then that could lead to a counting error. Once you put a task in `map_tasks` in `select_tasks_to_map` then you'll never see that task back in the list of `ready_tasks` ever again. It may go through `slice_task` and then new slices will appear in the `ready_tasks` of `select_tasks_to_map`. The one way you can see the same slice in `select_tasks_to_map` is if you relocate the slice to another mapper through `relocate_tasks` and then that other mapper sends it back to you (or you send it to yourself).

**Comment by bandokihiro** [comment]
Does this mean that currently, there isn't a way for me to make a difference in `select_tasks_to_map` between whether the task I see comes from after sharding or after `slice_task`? If that is the case, I don't see a way to prescribe correctly an order at the `select_tasks_to_map` stage in the pipeline to `map_task`.

**Comment by lightsighter** [comment]
I think the current answer is that is right. I think the solution will be for me to add an option to the `select_sharding_functor` mapper call for tasks that allows you to say that the resulting slice does not need to be recursively sliced. That way you won't have to run any tasks through `slice_task` unnecessarily.

**Comment by bandokihiro** [comment]
Cool, let me know then when you find time to add that option, thanks.

**Comment by lightsighter** [comment]
Try using this flag in `select_sharding_functor` to control whether to recurse the slicing of sharding output:

https://gitlab.com/StanfordLegion/legion/-/commit/3847f0c754caf269c337715c662fa327871d4fb3

**Comment by bandokihiro** [comment]
It works. I don't see multiple occurences in `select_tasks_to_map`. I then tried to implement what I originally wanted to do but I ran into some strange behavior (not immediately though, not sure why). Basically, it looks like Legion is ignoring task registration for the GPU. It tries to launch the OMP variant but then it seems that there is a hidden regression in the kokkos interop and it crashes. If I comment the OMP registration, then it complains that there is no variant for some tasks. I confirmed that the same app code (with or without the OMP registration commented) and input work on an old Legion without any issue.

**Comment by bandokihiro** [comment]
The OMP Kokkos interop doesn't work with `-ll:force_kthreads` which was responsible for the crash. But Legion still refuses to run the GPU variant.

**Comment by manopapad** [comment]
Is this related to the original issue? If not then please close this issue and create a new one, with the required context for the new problem.

**Comment by bandokihiro** [comment]
I haven't been able to get to where I want so I want to keep this one open. I can open another issue if the solution to my latest issue is not obvious.

**Comment by bandokihiro** [comment]
The two commits 3847f0c7 and 28ae84e4 are reponsible for this. According to @syamajala 1 rank/node is fine but Legion refuses to run tasks on the GPU for multiple ranks/node.

**Comment by lightsighter** [comment]
I don't think this is an issue with Legion. Both @bandokihiro and @syamajala are relying on `slice_task` as a crutch and that's not an option if you're going to set `slice_recurse=false` in the call to `select_sharding_functor`. You'll need to set the right destination processor in `select_task_options` because you're not going to get another chance to move the tasks before they map otherwise. I will switch the default back to `slice_recurse=true` for now to allow the crutch to still work, but if you want to set `slice_recurse=false` then you'll need to guarantee that you're already on the right processor for mapping. All the symptoms that you describe are because you're setting the wrong argument for `initial_proc` in `select_task_options` when also setting `slice_recurse=false`.

**Comment by bandokihiro** [comment]
From what I can tell, the `initial_proc` is set to a `TOC_PROC` (kind == 1). The mapper still fails within `default_find_preferred_variant` in `map_task`. Only a `TOC_PROC` variant is registered.

**Comment by syamajala** [comment]
In S3D at least we are setting initial_proc to a gpu in select_task_options. I could see that in the mapper log as well. In `map_task` though `task.target_proc.kind()` was always `LOC_PROC` when slice_recurse=false. 

https://gitlab.com/legion_s3d/legion_s3d/-/blob/regent_nscbc/rhst/rhst_mapper.cc#L163-175

**Comment by lightsighter** [comment]
I switched back the default so everything should be the same as it was before. You guys tell me what you want to do next.

**Comment by bandokihiro** [comment]
The `slice_recurse=true` path works (obviously) so my next step is to make the `slice_recurse=false` path work without trying to enforce an ordering in `select_tasks_to_map` yet (which would be my next next step). Currently, the `slice_recurse=false` path doesn't work with gpu variants not found in `map_task` and the `initial_proc` in `select_task_options` set correctly.

**Comment by lightsighter** [comment]
Pull and try again.

You guys still need to decide on what the default will be.

**Comment by bandokihiro** [comment]
Thanks, that fixes tasks not running on the GPU.

For the default behavior, I would advocate for `slice_recurse=true` in order not to break existing codes but that is just my personal opinion. 

Also since `select_sharding_functor` seems to not get called when there is only 1 rank, `slice_recurse` will take the default behavior for the 1-rank case. I haven't seen any problem with that but I thought it is worth mentioning.

For my next next goal, I am not able to hold onto a specific task in `select_tasks_to_map`. I have one mapper per rank. I log some information and the relevant bits are the following:
```
[0 - 20003dacf890]    2.845345 {3}{my_mapper}: In ready queue: evaluate_at_faces_task,
[0 - 20003dacf890]    2.845377 {3}{my_mapper}: Inserting evaluate_at_faces_task; unique ID = 1136
[0 - 20003e3cf890]    2.847842 {3}{my_mapper}: Mapping evaluate_at_faces_task; unique ID = 1138
[0 - 20003dacf890]    2.855024 {3}{my_mapper}: In ready queue: pack_for_explicit_ghosts_task,
[0 - 20003dacf890]    2.855065 {3}{my_mapper}: Inserting pack_for_explicit_ghosts_task; unique ID = 1152
[0 - 20003e84f890]    2.860467 {3}{my_mapper}: In ready queue: volume_residual_task,
[0 - 20003e84f890]    2.860499 {3}{my_mapper}: Holding on in select_tasks_to_map
[0 - 20003e3cf890]    2.875919 {3}{my_mapper}: Mapping pack_for_explicit_ghosts_task; unique ID = 1156
```
I want to wait for another task that depends on the result of `evaluate_at_faces_task` to be in the ready queue in order to let `volume_residual_task` to map. The app hangs after running `evaluate_at_faces_task` (the task completes) and I never see `pack_for_explicit_ghosts_task` launched. I looked at the backtraces but I haven't found anything interesting. I'll attach them just in case:
[gdb0.txt](https://github.com/StanfordLegion/legion/files/8376991/gdb0.txt)
[gdb1.txt](https://github.com/StanfordLegion/legion/files/8376992/gdb1.txt)
When I log `Holding on...`, I output an event that I wanted to trigger when I see `volume_residual_task` in `map_task`. It never gets to that point.

**Comment by bandokihiro** [comment]
nevermind, I didn't see an importance dependence which is the source of the hang

**Comment by bandokihiro** [comment]
`pack_for_explicit_ghosts_task` was in fact launched and run to completion. I still expected an update of the ready queue to trigger another call to `select_tasks_to_map` even if the event hasn't been triggered yet. The correct dependencies are as follows:
- `evaluate_at_faces_task` and `volume_residual_task` are immediately ready
- `pack_for_explicit_ghosts_task` depends on `evaluate_at_faces_task`
- `face_fluxes_task` depends on `pack_for_explicit_ghosts_task`

I wanted to put `volume_residual_task` and `face_fluxes_task` at the same time in the pipeline so that `volume_residual_task` runs after `pack_for_explicit_ghost_tasks`. Now I see that this approach might not be optimal because it creates a bubble in the pipeline. But I stilll expected the completion of `pack_for_explicit_ghosts_task` to trigger an update of the ready queue with now `face_fluxes_task` being ready and so I would see both `volume_residual_task` and `face_fluxes_task` in `select_tasks_to_map`. Instead, it hangs which might be expected behavior, I am not sure.

**Comment by bandokihiro** [comment]
My first attempt was dumm... I fixed it with the followings in `select_tasks_to_map`:
1. scan the ready queue
2. if `evaluate_at_faces_task` is in there, then give it a head start and return 
3. if a packing tasks is in there, then only allow them in the pipeline
4. otherwise revert the default logic

I think this achieves the behavior I want without playing with events.

**Comment by bandokihiro** [comment]
So I haven't been able to test it on a large proc count where I see sometimes an undesirable ordering but, looking at an old profile, I am worried this is not enough. In the attached pictures, both the yellow and blue tasks depend on the green task. I want the blue task to run after all yellow tasks. I assign higher mapping priority to the yellow tasks. Despite the fact that they are ready at the same time, the blue task sometimes go ahead. Note that this picture was obtained with the default logic in `select_tasks_to_map` but I don't see how my above changes would help in that situation. Would you agree?
![Screenshot from 2022-03-30 22-14-15](https://user-images.githubusercontent.com/39278362/161099085-8ad1f2ec-8779-409b-9661-aa4167d53a83.png)

**Comment by bandokihiro** [comment]
Summarizing here what I ended up doing before closing

- logic within `select_tasks_to_map` is not going to work in a trace replay since this method doesn't get called
- I chose to create a write-after-read dummy region dependency between all yellow tasks and the blue task such that the latter runs after the former, this will work in a trace replay as long as the yellow tasks read from the same instance as the blue task is going to write to (they have to be issued first of course)
- there was a suggestion to make realm respect mapping priorities for independent operations, see the associated bullet point in #407
